Security Considerations (#441)
decentralized-identity/presentation-exchange: spec/spec.md

#### Impact of not using function extensions in JSONPath
As such it is recommended that software implementors do not allow for function extensions in JSONPath and relying parties do not create definitions that
rely on functions.

Not using functions does restricts the input descriptor part in terms of flexibility. 
On the other hand, most issuers know exactly what type of credentials and claims they want to receive, 
so having very complex matching logic with Regular Expressions involved is not typically needed.

#### Restrict submission data to exact JSONPath paths
JSON path has filtering logic, but once you are submitting your Verifiable Presentation and submission data, you 
should be correlating input descriptor field ids in the `descriptor_map` with exact paths to the Verifiable Credentials 
or claims. We suggest to not allow for both filters and function extensions for parties implementing the specification.

#### Attack Vectors on JSON Schema
JSON Schema validation allows the use of Regular Expressions for instance in filters, to which the security 
considerations, mentioned above, apply. Regular Expressions could be used in the `filter` of a `field` as part

> of a `constraint`, by using the `pattern` property. We advise not to use this property.

Reviewer: We'll also want to exclude the `regex` format keyword https://json-schema.org/draft/2020-12/json-schema-validation.html#section-7.3.8

and `patternProperties` https://json-schema.org/draft/2020-12/json-schema-core.html#section-10.3.2.2

## Appendix

### Security Considerations
JSONPath and JSON schemas are being used in this specification. Both allow the use of Regular Expressions, which have 

> numerous different (often incompatible) implementations. Some implementations allow the embedding of arbitrary code, 
> which MUST never be permitted, given the presentation definition and submission data should be treated as input 
> from untrusted sources. 

Reviewer 2: ```suggestion
numerous different (often incompatible) implementations. The embedding of arbitrary code in Regular Expressions MUST never be permitted, given the presentation definition and submission data should be treated as input 
from untrusted sources. 
```

Regular expressions can often also be crafted to be extremely expensive to compute, resulting in
denial-of-service attacks.

As such this specification recommends not using Regular Expressions at all in your implementation when using JSONPath 
and JSON Schemas, at the potential expense of expressiveness and/or usability. If you really want to use them, make sure
the code is properly guarded both in terms of execution time and from a security perspective, by for instance 

> executing the code dealing with the regular expressions in a sandbox. 

Reviewer 2: why not prohibit using it at all? what expressiveness/usability benefits are we aware of?

Author: Good question. It becomes harder to create certain more broad definitions. But as I argued in the PE meetings, I don't think that it would be good to say anything like 

'Just give me X, Y or Z, I don't really care where it comes from or what other constraints are present'

Normally a RP exactly knows that types of claims and Credentials it is requesting. You still have things like enumerations and constants etc. So IMO we could ditch using regexes entirely.

When allowing Regular Expression Implementers need to be aware that good average performance is not sufficient as long
as an attacker can choose to submit specially crafted Regular Expressions that can result in surprisingly high, possibly

> exponential, CPU usage or for example result in a stack overflow. Also when you do allow Regular Expressions in your
> solution, for instance in the Presentation Definition, you will risk the other side not supporting it, as they took the
>  cautious route of not allowing Regular Expressions. This specification is still suggesting to not allow Regular 

Reviewer 2: ```suggestion
exponential, CPU usage or for example result in a stack overflow. Also when Regular Expressions are allowed in the solution, for instance in the Presentation Definition, there is the risk of the other side not supporting it, being cautious. This specification still suggests not to allow Regular 
```

> Expressions, to not break backwards compatibility with version 2.0.0. A future new major version is highly likely to 

Reviewer 2: which parts of PEv2 depend on RegEx..?

Author: JSONPath function extensions (part of the JSONPath spec) and filters in fields, as used in several examples in the spec (which should be constants anyway btw)

not allow for usage of Regular Expressions in a normative way.

#### JSONPath

JSONPath is being used in this specification both in the Input Descriptors of the Presentation Definition,

> defining where certain data is to be found, and the Submission Data,

Reviewer 2: ```suggestion
defining where certain data is to be found, and the Presentation Submission,
```

defining how data maps from the definition to the Verifiable Presentation.
With Regular Expressions come certain security considerations, that must be taken into account.
Be aware that in a future major version of the spec we envision to have some normative texts in the specification

> itself disallowing certain aspects of JSONPath altogether like function extensions (ie regular expressions)

Reviewer 2: ```suggestion
itself disallowing certain aspects of JSONPath altogether like function extensions (i.e., Regular Expressions)
```

> Please also read the JSONPath IETF document providing more details on some of the security considerations mentioned

Reviewer 2: pointing to a concrete url would be nice https://datatracker.ietf.org/api/v1/doc/document/draft-ietf-jsonpath-base/

Author: Yeah thanks, still need to do those. This is just a draft to solicit some feedback

below.

#### Attack Vectors on JSONPath Implementations

Historically, JSONPath has often been implemented by feeding parts of the query to an underlying 
programming language engine, e.g.,  JavaScript's eval() function.  This approach is well known to lead to 
injection attacks and would require perfect input validation to prevent these attacks. 
Since that is hard to achieve we recommend restricting JSONPath queries to not allow for functions 

> (i.e Regular Expressions, see above)

Reviewer 2: ```suggestion
(i.e., Regular Expressions, see above)
```

Attacks on availability may attempt to trigger unusually expensive runtime performance exhibited by certain 
implementations in certain cases. This particularly applies to the use of Regular Expressions. We recommend restricting 
JSONPath queries to not allow for function extensions.
Function extensions like `match` and `search` are using regular expressions, 
to which the above security considerations apply. 

#### Impact of not using function extensions in JSONPath

> As such we suggest software implementors to not allow for functions and relying parties to no create definitions that 

Reviewer 2: ```suggestion
As such it is recommended that software implementors do not allow for function extensions in JSONPath and relying parties do not create definitions that
```

rely on functions.

> Not using functions does restrict the input descriptor part in terms of how flexible you can be, 

Reviewer 2: ```suggestion
Not using functions does restricts the input descriptor part in terms of how flexible you can be, 
```

> Not using functions does restrict the input descriptor part in terms of how flexible you can be, 
> on the other hand most issuers know exactly what type of credentials and claims they want to receive anyway, 
> so having very complex matching logic with Regular Expressions involved typically isn’t needed anyway.

Reviewer 2: ```suggestion
Not using functions does restricts the input descriptor part in terms of flexibility. 
On the other hand, most issuers know exactly what type of credentials and claims they want to receive, 
so having very complex matching logic with Regular Expressions involved is not typically needed.
```

### Developer Resources

> #### JSON Schema implementations

Reviewer: If we're excluding JsonPath implementations, shouldn't we exclude JSON Schema impls as well? If not, should we instead link to the official implementation list? https://json-schema.org/implementations.html
